cxx-idioms/named_params_2.cpp

```diff
@@ -0,0 +1,57 @@
+#include <iostream>
+#include <string>
+#include <tuple>
+
+template <class T>
+struct make_param
+{
+  make_param(T&& o): v(std::move(o)) { }
+  make_param(T const& o): v(o) { }
+  operator T() { return std::move(v); }
+private:
+  T v;
+};
+
+#define DECL_PARAM(NAME, TYPE)\
+  struct NAME: make_param<TYPE> { using make_param<TYPE>::make_param; }
+
+struct robot {
+
+  struct param
+  {
+    DECL_PARAM(name, std::string);
+    DECL_PARAM(age, size_t);
+    DECL_PARAM(weight, size_t);
+  };
+
+  template<class... Params>
+  robot(Params&& ... p)
+  {
+    std::initializer_list<int> v{(apply(std::forward<Params>(p)), 1)...};
```

Comment:
like in the other one right? the issue with the original deign is that's you can't have two member objects with the same type! (that rings a bell to something maybe ;))